src/python/pants/backend/python/rules/repl.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from pathlib import Path
+
+from pants.backend.python.rules.pex import Pex
+from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
+from pants.base.build_root import BuildRoot
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
+from pants.engine.legacy.graph import HydratedTargets
+from pants.engine.legacy.structs import PythonTargetAdaptor
+from pants.engine.rules import goal_rule
+from pants.engine.selectors import Get
+from pants.option.global_options import GlobalOptions
+from pants.util.contextutil import temporary_dir
+
+
+logger = logging.getLogger(__name__)
+
+
+class ReplOptions(GoalSubsystem):
+  """Opens a REPL."""
+  name = 'repl2-python'
+
+
+class Repl(Goal):
+  subsystem_cls = ReplOptions
+
+
+@goal_rule
+async def run_python_repl(
+    console: Console,
+    workspace: Workspace,
+    runner: InteractiveRunner,
+    build_file_addresses: BuildFileAddresses,
+    build_root: BuildRoot,
+    global_options: GlobalOptions) -> Repl:
+
+
+  hydrated_targets = await Get[HydratedTargets](BuildFileAddresses, build_file_addresses)
```

Comment:
Instead, you can directly request `HydratedTargets` in the rule. It looks like you aren't using `build_file_addresses` anywhere else.

Reply:
Also, I think we need `TransitiveHydratedTargets`?